### PR TITLE
load_mon module for Linux platform

### PR DESCRIPTION
**Pull request by SalimTerryLi**
This new module publishes cpuload message over uORB, to satisfy the pre-arm check on Linux platforms, especially on RaspBerry Pi.

The module get CPU usage of PX4 process by calling [times()](https://linux.die.net/man/2/times).

Test data:
```
pxh> listener cpuload

TOPIC: cpuload
 cpuload_s
	timestamp: 9270820566  (0.277749 seconds ago)
	load: 0.4200
	ram_usage: 0.1067
pxh> listener cpuload

TOPIC: cpuload
 cpuload_s
	timestamp: 9271320556  (0.139869 seconds ago)
	load: 0.3800
	ram_usage: 0.1067

```
Or, should RPi simply bypass this check?

### CODE REVIEW

**Comment by SalimTerryLi** [comment]
Found #15330. Is there any plan to implement that framework on Linux? Or any ideas?

**Comment by dagar** [comment]
> Found #15330. Is there any plan to implement that framework on Linux? Or any ideas?

That only applies to PX4/NuttX where calculating the load (including what's displayed in top) is handled PX4 side via scheduling hooks in NuttX.

**Comment by dagar** [comment]
Looks good @SalimTerryLi. In this case the small chunk of logic difference between NuttX and Linux is pretty small. I wouldn't be opposed to carrying this with defines in the existing `load_mon`. What do you think?

https://github.com/PX4/Firmware/blob/258e5ca5c986cc39e4e656dd38c071f10180e5cf/src/modules/load_mon/LoadMon.cpp#L117-L123

**Comment by dagar** [comment]
> Or, should RPi simply bypass this check?

Might as well have it respect the check if the data (`cpu_load`) is available.

**Comment by SalimTerryLi** [comment]
> 
> 
> Looks good @SalimTerryLi. In this case the small chunk of logic difference between NuttX and Linux is pretty small. I wouldn't be opposed to carrying this with defines in the existing `load_mon`. What do you think?
> 
> https://github.com/PX4/Firmware/blob/258e5ca5c986cc39e4e656dd38c071f10180e5cf/src/modules/load_mon/LoadMon.cpp#L117-L123

That may need a bunch of ifdefs, but is able to optimize. I agree too. I have a question that [this](https://github.com/PX4/Firmware/blob/pr-cpuload_updates/platforms/posix/src/px4/common/cpuload.cpp) also exists in POSIX framework. Should we implement cpuinfo here and make another wrapper which can be shared with Nuttx? Or just leave it here and get CPU data inside the module?

**Comment by dagar** [comment]
> I have a question that this also exists in POSIX framework. Should we implement cpuinfo here and make another wrapper which can be shared with Nuttx? Or just leave it here and get CPU data inside the module?

It looks like that only exists for a simple `top` within PX4 when running on MacOS.

https://github.com/PX4/Firmware/blob/258e5ca5c986cc39e4e656dd38c071f10180e5cf/src/lib/systemlib/print_load_posix.c#L97-L127

The `cpuload` topic is mildly useful because it's logged throughout flight and integrated with commander (preflight checks and lights). In logs actual `top` output is captured once at arming (preflight data) and once again when disarming (postflight data). 

So not having this on Linux isn't critical, but if you wanted it I would consider if grabbing system `top` output is an easy alternative rather than PX4 absorbing functionality it doesn't really benefit from.

**Comment by SalimTerryLi** [comment]
Can not enable it in SITL because I don't know how to implement it with OSX. I'm also interested in implementing top command on Linux, grabbing information from `/proc/self/task`.

File: src/modules/load_mon/LoadMon.cpp [hunk 1]
```diff
@@ -211,6 +246,8 @@ usage and publish the `cpuload` topic.
 
 On NuttX it also checks the stack usage of each process and if it falls below 300 bytes, a warning is output,
 which will also appear in the log file.
+
+On Linux stack check is disabled.
```
Comment (dagar) [inline src/modules/load_mon/LoadMon.cpp:250]: Probably better to say it's only support on NuttX (there's also MacOS and QuRT).

File: src/modules/load_mon/LoadMon.hpp [hunk 2]
```diff
@@ -74,16 +79,24 @@ class LoadMon : public ModuleBase<LoadMon>, public ModuleParams, public px4::Sch
 	/** Do a calculation of the CPU load and publish it. */
 	void cpuload();
 
+	/* Don't check stack usage on Linux */
+#ifndef __PX4_LINUX
```
Comment (dagar) [inline src/modules/load_mon/LoadMon.hpp:83]:
```suggestion
#if defined(__PX4_NUTTX)
```

File: src/modules/load_mon/LoadMon.hpp [hunk 3]
```diff
@@ -74,16 +79,24 @@ class LoadMon : public ModuleBase<LoadMon>, public ModuleParams, public px4::Sch
...
 	uORB::PublicationQueued<task_stack_info_s> _task_stack_info_pub{ORB_ID(task_stack_info)};
-	uORB::Publication<cpuload_s> _cpuload_pub{ORB_ID(cpuload)};
-
-	hrt_abstime _last_idle_time{0};
+#endif
+	uORB::Publication<cpuload_s> _cpuload_pub {ORB_ID(cpuload)};
+
+#ifdef __PX4_LINUX
+	/* calculate usage directly from clock ticks on Linux */
+	clock_t _last_total_time_stamp = 0, _last_spent_time_stamp = 0;
```
Comment (dagar) [inline src/modules/load_mon/LoadMon.hpp:95]:
```suggestion
	clock_t _last_total_time_stamp{};
	clock_t _last_spent_time_stamp{};
```

File: src/modules/load_mon/LoadMon.cpp [hunk 4]
```diff
@@ -102,6 +108,23 @@ void LoadMon::Run()
...
+		// Just get the time in the first iteration */
+		_last_total_time_stamp = total_time_stamp;
+		_last_spent_time_stamp = spent_time_stamp;
+		return;
+	}
+
+	// compute system load
+	const float interval = total_time_stamp - _last_total_time_stamp;
+	const float interval_spent_time = spent_time_stamp - _last_spent_time_stamp;
+#else
```
Comment (dagar) [inline src/modules/load_mon/LoadMon.cpp:126]:
```suggestion
#elif defined(__PX4_NUTTX)
```

File: src/modules/load_mon/LoadMon.cpp [hunk 5]
```diff
@@ -88,10 +90,14 @@ void LoadMon::Run()
 
 	cpuload();
 
+#ifndef __PX4_LINUX
```
Comment (dagar) [inline src/modules/load_mon/LoadMon.cpp:93]:
```suggestion
#if defined(__PX4_NUTTX)
```

File: src/modules/load_mon/LoadMon.cpp [hunk 6]
```diff
@@ -33,6 +33,7 @@
 
 #include "LoadMon.hpp"
 
+#ifndef __PX4_LINUX
```
Comment (dagar) [inline src/modules/load_mon/LoadMon.cpp:36]:
```suggestion
#if defined(__PX4_NUTTX)
```

**Comment by dagar** [comment]
Looks good @SalimTerryLi. I would just change the defines to be focused around if NuttX or if Linux rather than ifndef to avoid potential confusion later with MacOS and QuRT.

File: src/modules/load_mon/LoadMon.cpp [hunk 7]
```diff
@@ -41,6 +42,7 @@ static constexpr unsigned STACK_LOW_WARNING_THRESHOLD = 300;
 #endif
 
 static constexpr unsigned FDS_LOW_WARNING_THRESHOLD = 2; ///< if free file descriptors fall below this, print a warning
+#endif // __PX4_LINUX, disable stack related issue
```
Comment (dagar) [inline src/modules/load_mon/LoadMon.cpp:45]:
```suggestion
#endif // __PX4_NUTTX
```

File: src/modules/load_mon/LoadMon.hpp [hunk 8]
```diff
@@ -74,16 +79,25 @@ class LoadMon : public ModuleBase<LoadMon>, public ModuleParams, public px4::Sch
...
-
-	hrt_abstime _last_idle_time{0};
+#endif
+	uORB::Publication<cpuload_s> _cpuload_pub {ORB_ID(cpuload)};
+
+#ifdef __PX4_LINUX
+	/* calculate usage directly from clock ticks on Linux */
+	clock_t _last_total_time_stamp{};
+	clock_t _last_spent_time_stamp{};
+#else
```
Comment (dagar) [inline src/modules/load_mon/LoadMon.hpp:97]:
```suggestion
#elif defined(__PX4_NUTTX)
```

**Comment by dagar** [comment]
Add a `load_mon start` in the Linux init scripts? 

https://github.com/PX4/Firmware/blob/master/posix-configs/rpi/px4.config#L28

Which reminds me that we really need to merge Linux init with the rest of PX4 (https://github.com/PX4/Firmware/tree/master/ROMFS/px4fmu_common).

**Pull request merged**